### PR TITLE
[Enhancement] Supports struct and null arguments in function expressions

### PR DESCRIPTION
**Pull request by zenoyang**
Why I'm doing:
An error is reported when the `struct` type is used with `null` as an argument to the `if` function.

```sql
MySQL [ssb]> select count(distinct if(lo_quantity > 10, row(lo_orderkey, lo_linenumber), row(lo_orderkey, lo_custkey))) from lineorder;
+-----------------------------------------------------------------------------------------------------+
| count(DISTINCT if(lo_quantity > 10, row(lo_orderkey, lo_linenumber), row(lo_orderkey, lo_custkey))) |
+-----------------------------------------------------------------------------------------------------+
|                                                                                            56772832 |
+-----------------------------------------------------------------------------------------------------+
1 row in set (3.20 sec)

MySQL [ssb]> select count(distinct if(lo_quantity > 10, row(lo_orderkey, lo_linenumber), null)) from lineorder;
ERROR 1064 (HY000): Getting analyzing error from line 1, column 22 to line 1, column 92. Detail message: No matching function with signature: if(boolean, struct<col1 int(11), col2 int(11)>, NULL_TYPE).
```
After repair:
```sql
MySQL [ssb]> select count(distinct if(lo_quantity > 10, row(lo_orderkey, lo_linenumber), null)) from lineorder;
+-----------------------------------------------------------------------------+
| count(DISTINCT if(lo_quantity > 10, row(lo_orderkey, lo_linenumber), NULL)) |
+-----------------------------------------------------------------------------+
|                                                                    48465645 |
+-----------------------------------------------------------------------------+
1 row in set (3.36 sec)
```

What I'm doing:
In the `resolveArgTypes` method, `fn.getArgs()` if the type is `AnyStructType` and `inputArgTypes` is `NULL_TYPE`, `resolvedTypes` will be converted to `StructType(boolean)`. This causes the struct common type to return `type.invalid`, resulting in an error.
So I modified the struct common type logic, regardless of the length of the struct, to support `null` scenarios.

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/common/TypeManager.java [hunk 1]
```diff
@@ -104,18 +104,25 @@ private static Type getCommonMapType(MapType t1, MapType t2) {
     }
 
     private static Type getCommonStructType(StructType t1, StructType t2) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/common/TypeManager.java:106]: the method will modify cast logical, I think should need describe `what are you doing` and add more test case @wangsimo0 @yingtingdong struct common type change.
the common type of `struct<a int, b bigint>` and `struct<a bigint, b int, c int>` is `struct<a bigint, b bigint, c int>` is

Comment (zenoyang) [reply fe/fe-core/src/main/java/com/starrocks/sql/common/TypeManager.java:106]: @Seaven Thanks, I updated `What I'm doing`

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/common/TypeManager.java:106]: I get your intention, but I don’t recommend changing the code `getCommonStructType`, the method will effect `cast(row(1,2) as struct<a int, b int, c int>)`, and BE doesn't support now.

If you only want to fix `IF(xx, NULL, struct)` bug, I think you can handle it in `PolymorphicFunctionAnalyzer.resolveByDeducingReturnType`, to handle NULL by a speical process

Comment (zenoyang) [reply fe/fe-core/src/main/java/com/starrocks/sql/common/TypeManager.java:106]: @Seaven Good suggestion, I will optimize it again.

**Comment by before-Sunrise** [comment]
fixed by https://github.com/StarRocks/starrocks/pull/39484

**Pull request closed without merging**